File: src/client/java/net/legitimoose/bot/World.java
package net.legitimoose.bot;

import com.google.gson.JsonObject;
import com.mongodb.client.MongoCollection;
import com.mongodb.client.MongoDatabase;
import com.mongodb.client.model.UpdateOptions;
import com.mongodb.client.model.Updates;
import org.bson.BsonArray;
import org.bson.Document;
import org.bson.conversions.Bson;
import org.jetbrains.annotations.ApiStatus;

import static com.mongodb.client.model.Filters.eq;
import static com.mongodb.client.model.Filters.lt;
import static net.legitimoose.bot.LegitimooseBot.LOGGER;

public record World(
        String creation_date,
        int creation_date_unix_seconds,
        boolean enforce_whitelist,
        boolean locked,
        String owner_uuid,
        int player_count,
        int max_players,
        String resource_pack_url,
        String world_uuid,
        String version,
        int visits,
        int votes,
        boolean whitelist_on_version_change,
        String name,
        String description,
        String raw_name,
        String raw_description,
        int featured_instant,

        @Deprecated
        @ApiStatus.ScheduledForRemoval(inVersion = "3.0.0") // API v4
        boolean jam_world,

        @Deprecated
        @ApiStatus.ScheduledForRemoval(inVersion = "3.0.0") // API v4
        int jam_id,

        JsonObject jam,

        String icon,
        long last_scraped
) {
    public void write(MongoDatabase db) {
        MongoCollection<World> coll = db.getCollection("worlds", World.class);

        coll.deleteMany(lt("last_scraped", System.currentTimeMillis() / 1000L - 86400));

        LOGGER.info("writing world");
        Bson updates =
                Updates.combine(
                        Updates.set("creation_date", this.creation_date),
                        Updates.set("creation_date_unix_seconds", this.creation_date_unix_seconds),
                        Updates.set("enforce_whitelist", this.enforce_whitelist),
                        Updates.set("locked", this.locked),
                        Updates.set("owner_uuid", this.owner_uuid),
                        Updates.set("player_count", this.player_count),
                        Updates.set("max_players", this.max_players),
                        Updates.set("resource_pack_url", this.resource_pack_url),
                        Updates.set("version", this.version),
                        Updates.set("visits", this.visits),
                        Updates.set("votes", this.votes),
                        Updates.set("whitelist_on_version_change", this.whitelist_on_version_change),
                        Updates.set("name", this.name),
                        Updates.set("description", this.description),
                        Updates.set("raw_name", Document.parse(this.raw_name)),
                        Updates.set("raw_description", BsonArray.parse(this.raw_description)),
                        Updates.set("featured_instant", this.featured_instant),
                        Updates.set("jam_world", this.jam_world),
                        Updates.set("jam_id", this.jam_id),
                        Updates.set("jam", Document.parse(this.jam.toString())),
                        Updates.set("icon", this.icon),
                        Updates.set("last_scraped", this.last_scraped));
        coll.updateOne(eq("world_uuid", this.world_uuid), updates, new UpdateOptions().upsert(true));
        LOGGER.info("Wrote world {}", this.world_uuid);
    }
}

File: src/client/kotlin/net/legitimoose/bot/World.kt
package net.legitimoose.bot

import com.mongodb.MongoClientException
import com.mongodb.client.model.Filters.eq
import com.mongodb.client.model.Filters.lt
import com.mongodb.client.model.UpdateOptions
import com.mongodb.client.model.Updates
import com.mongodb.kotlin.client.MongoCollection
import com.mongodb.kotlin.client.MongoDatabase
import net.legitimoose.bot.LegitimooseBot.logger
import org.bson.Document
import org.bson.types.ObjectId

data class World(
    val creation_date: String,
    val creation_date_unix_seconds: Int,
    val enforce_whitelist: Boolean,
    val locked: Boolean,
    val owner_uuid: String,
    val player_count: Int,
    val resource_pack_url: String,
    val world_uuid: String,
    val version: String,
    val visits: Int,
    val votes: Int,
    val whitelist_on_version_change: Boolean,
    val name: String,
    val description: String,
    val raw_name: String,
    val raw_description: String,
    val icon: String,
    val last_scraped: Long
) {
  fun upload(db: MongoDatabase) {
    val coll: MongoCollection<Document> = db.getCollection("worlds")
    var doc: Document?
    try {
      doc = coll.find(eq("world_uuid", this.world_uuid)).first()
    } catch (e: MongoClientException) {
      doc = null
    }
    coll.deleteMany(lt("last_scraped", System.currentTimeMillis() / 1000L - 86400))
    if (doc != null) {
      logger.info("updating world")
      val updates =
          Updates.combine(
              Updates.set("enforce_whitelist", this.enforce_whitelist),
              Updates.set("locked", this.locked),
              Updates.set("owner_uuid", this.owner_uuid),
              Updates.set("player_count", this.player_count),
              Updates.set("resource_pack_url", this.resource_pack_url),
              Updates.set("version", this.version),
              Updates.set("visits", this.visits),
              Updates.set("votes", this.votes),
              Updates.set("whitelist_on_version_change", this.whitelist_on_version_change),
              Updates.set("name", this.name),
              Updates.set("description", this.description),
              Updates.set("raw_name", this.raw_name),
              Updates.set("raw_description", this.raw_description),
              Updates.set("icon", this.icon),
              Updates.set("last_scraped", this.last_scraped))
      coll.updateOne(eq("world_uuid", this.world_uuid), updates, UpdateOptions())
      logger.info("Updated world")
      return
    }
    coll.insertOne(
        Document()
            .append("_id", ObjectId())
            .append("creation_date", this.creation_date)
            .append("creation_date_unix_seconds", this.creation_date_unix_seconds)
            .append("enforce_whitelist", this.enforce_whitelist)
            .append("locked", this.locked)
            .append("owner_uuid", this.owner_uuid)
            .append("player_count", this.player_count)
            .append("resource_pack_url", this.resource_pack_url)
            .append("world_uuid", this.world_uuid)
            .append("version", this.version)
            .append("visits", this.visits)
            .append("votes", this.votes)
            .append("whitelist_on_version_change", this.whitelist_on_version_change)
            .append("name", this.name)
            .append("description", this.description)
            .append("raw_name", this.raw_name)
            .append("raw_description", this.raw_description)
            .append("icon", this.icon)
            .append("last_scraped", this.last_scraped))
    logger.info("Created world")
  }
}
